File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20141229_Running_Time_of_Algorithms/Solution.java
import java.util.Scanner;

public class Solution {
  public static int insertionSort(int[] arr) {
    int cnt = 0;
    for ( int i = 1; i < arr.length; ++i )  {
      for ( int j = i; j > 0; --j ) {
        if ( arr[j] < arr[j - 1] )  {
          int tmp = arr[j];
          arr[j] = arr[j - 1];
          arr[j - 1] = tmp;
          ++cnt;
        }
      }
    }
    return cnt;
  }

  public static void main(final String[] args)  {
    Scanner sc = new Scanner(System.in);
    int s = Integer.parseInt(sc.nextLine());
    String[] inps = sc.nextLine().split(" ");
    int[] arr = new int[inps.length];
    for ( int i = 0; i < arr.length; ++i )
      arr[i] = Integer.parseInt(inps[i]);
    System.out.println(insertionSort(arr));
  }
}

File: Problems/hacker_rank/Algorithm/Arrays_and_Sorting/20141229_Running_Time_of_Algorithms/Solution.scala
object Solution {
  def insertionSort(arr: Array[Int]): Int = {
    var cnt: Int = 0
    for ( i <- 1 until arr.length ) {
      for ( j <- i until 0 by -1 )  {
        if ( arr(j) < arr(j - 1) )  {
          //  http://stackoverflow.com/questions/10158405/swapping-array-values-with-for-and-yield-scala
          val tmp = arr(j)
          arr(j) = arr(j - 1)
          arr(j - 1) = tmp
          cnt += 1
        }
      }
    }
    cnt
  }

  def main(args: Array[String]) {
    val inps = io.Source.stdin.getLines.drop(1).toArray
    val arr = inps(0).split(" ").map((s: String) => s.toInt).toArray
    println(insertionSort(arr))
  }
}
